scripts/usb_scan.py

```diff
@@ -0,0 +1,638 @@
+
+    def add_interface(self, interface):
+        """ add an interface to this device
+
+        :param interface:(UsbInterface) the UsbInterface to add
+        :return: None
+        """
+        if interface in self.interfaces:
+            log.debug("overriding existing interface: " + interface)
+            self.interfaces.remove(interface)
+        self.interfaces.add(interface)
+
+    def del_interface(self, interface):
+        """remove an interface from this device
+
+        :param interface:(UsbInterface) the UsbInterface to remove
+        :return: None
+        """
+        if interface in self.interfaces:
+            log.debug("removing interface: " + interface)
+            self.interfaces.remove(interface)
+
+    def get_all_interfaces(self):
+        """ get all interfaces attached of this device
+
+        :return: set of all interfaces
+        """
+        return self.interfaces
+
+    def is_ready(self):
+        """ check if this device has all the interfaces attached
+
+        :return: bool, if it's ready to do policy check now
+        """
+        n = int(self[self._NUM_INTERFACES])
+        return n > 0 and n == len(self.interfaces)
+
+
+class UsbInterface(UsbObject):
+    """ Class for USB interface, save USB properties in UsbObject dict
+
+    """
+    _NUMBER = "bInterfaceNumber"
+    _CLASS = "bInterfaceClass"
+    _SUB_CLASS = "bInterfaceSubClass"
+    _PROTOCOL = "bInterfaceProtocol"
+
+    _PROPS = [_NUMBER, _CLASS, _SUB_CLASS, _PROTOCOL]
+
+    def __init__(self, node, props):
+        """ initialise UsbInterface, set node and properties
+
+        :param node(str): device node
+        :param props(pyudev.Device.attributes): device attributes, to get
+        properties from sysfs
+        """
+        super(UsbInterface, self).__init__(node)
+        for p in self._PROPS:
+            if props.get(p) is not None:
+                self[p] = props.get(p)
+
+    def debug_str(self, level=0):
+        s = super(UsbInterface, self).debug_str(level)
+        return s
+
+    def is_class_hub(self):
+        return self._is_class_hub(self._CLASS)
+
+    def is_initialized(self):
+        # usb_intf_attr(bInterfaceNumber, "%02x\n");
+        # usb_intf_attr(bInterfaceClass, "%02x\n");
+        # usb_intf_attr(bInterfaceSubClass, "%02x\n");
+        # usb_intf_attr(bInterfaceProtocol, "%02x\n");
+        for p in self._PROPS:
+            if p not in self or not self.validate_int(self[p], 16):
+                return False
+        return super(UsbInterface, self).is_initialized()
+
+    def is_child_of(self, parent):
+        if isinstance(parent, UsbDevice) and parent.is_initialized():
+            conf_value = parent[UsbDevice._CONF_VALUE]
+            pattern = r"^{}:{}\.\d+$".format(re.escape(parent.get_node()),
+                                             re.escape(conf_value))
+            return re.match(pattern, self.get_node()) is not None
+        return False
+
+
+def get_usb_info():
```

Comment:
If `pyudev.Context` or `context.list_devices` are failing, this script will fail.
I think this can be fixed in a separate PR after having investigated if and how thay can fail